### ISSUE TITLE
Compiler error "Unable to generalize a type variable. It is not unranked."

### ISSUE THREAD

**Issue by mcapodici** [elm-lang/elm-compiler#1681]
I got the following message when building my project

>  - elm-make.exe: It looks like something went wrong with the type inference algorithm.
> 
> Unable to generalize a type variable. It is not unranked.
> 
> Please create a minimal example that triggers this problem and report it to
> <https://github.com/elm-lang/elm-compiler/issues>
> elm-make.exe: thread blocked indefinitely in an MVar operation
> 

Version Numbers:
  OS: Windows 10
  Elm: 0.18.0
  Browser: N/A - compile time issue

Below is as minimal as I can get the example, in one file, it stopped happening when I started to reduce the applyEditingAccountRow any further:

    import Json.Decode exposing (..)
    import Json.Encode as JsonE
    import Task exposing (Task)
    import Date exposing (Date)

    attempt2 : (x->msg)->(a->msg)-> Task x a -> Cmd msg
    attempt2 errorMap successMap  =
      Task.attempt (\result ->
        case result of
          Ok r -> successMap r
          Err e -> errorMap e) 

    type alias Account =
        { storedKey : StoredKey
        , name : String
        }

    type alias EditingAccountRow =
        { storedKey : StoredKey
        , name : String
        }

    type alias AccountPageModel =
        { editingAccountRow : Maybe EditingAccountRow
        , confirmDeleteKey : Maybe StoredKey
        , nextFakeId : Int
        }

    updateItemWithKey : WithStoredKey a -> StoredKey -> a -> List a -> List a
    updateItemWithKey dict key newItem list =
        List.map
            (\item ->
                if dict.get item == key then
                    newItem
                else
                    item
            )
            list

    withStoredKey_EditingAccountRow : WithStoredKey EditingAccountRow
    withStoredKey_EditingAccountRow =
        { get = .storedKey, set = \key acc -> { acc | storedKey = key } }

    withStoredKey_Account : WithStoredKey Account
    withStoredKey_Account =
        { get = .storedKey, set = \key acc -> { acc | storedKey = key } }

    withLocalId_AccountPageModel : WithLocalId AccountPageModel
    withLocalId_AccountPageModel =
        { get = .nextFakeId, set = \id model -> { model | nextFakeId = id } }

    toFromEditable_Account : ToFromEditable Account EditingAccountRow
    toFromEditable_Account =
        { toEditable = \data -> { storedKey = data.storedKey, name = data.name }
        , toData = \row -> { storedKey = row.storedKey, name = row.name }
        }

    type alias ToFromEditable data editable =
        { toEditable : data -> editable, toData : editable -> data }

    type alias WithStoredKey a =
        { get : a -> StoredKey, set : StoredKey -> a -> a }

    type alias WithLocalId a =
        { get : a -> Int, set : Int -> a -> a }

    type alias Persister a =
        { encode : a -> JsonE.Value
        , decoder : Decoder a
        , className : String
        }

    type alias ErrorMessage =
        String

    type alias StoredKey =
        String

    type alias Clause =
        { fieldName : String
        , value : JsonE.Value
        }

    type alias Save a =
        a -> Task ErrorMessage StoredKey

    type alias Get a =
        List Clause -> Task ErrorMessage (List a)

    type alias Update a =
        StoredKey -> a -> Task ErrorMessage ()

    type alias Delete =
        StoredKey -> Task ErrorMessage ()

    type alias BaaSPersister a =
        { save : Save a
        , get : Get a
        , update : Update a
        , delete : Delete
        }

    applyEditingAccountRow : ToFromEditable account editAccount -> WithStoredKey account -> WithStoredKey editAccount -> WithLocalId AccountPageModel -> AccountPageModel -> BaaSPersister account -> editAccount -> List account -> ( List account, Cmd Msg, AccountPageModel )
    applyEditingAccountRow toFromEditableDict withStoredKeyDict withStoredKeyEditDict withLocalIdDict model persister row accounts =
        let editRowStoredKey = withStoredKeyEditDict.get row 
            updatedAccount = toFromEditableDict.toData row
        in
                ( updateItemWithKey withStoredKey_Account editRowStoredKey updatedAccount accounts
                , persister.update editRowStoredKey updatedAccount
                    |> attempt2 StorageFailed (always StorageUpdateSuccess)
                , model
                )

    type
        Msg
        {--Button Actions --}
        = AddAccount
        | SaveAccount
        | CancelAccount
        | EditAccount StoredKey
        | DeleteAccount StoredKey Bool
        | CancelDelete
        | DoubleClickAccount StoredKey String
          {--Keystrokes --}
        | SetAccountName String
          {--API--}
        | StorageUpdateSuccess
        | StorageDeleteSuccess
        | StorageSaveSuccess StoredKey StoredKey
        | StorageFailed String

**Comment by zwilias** [comment]
This error generally shows up when the code places constraints on the type variables that don't match the signature. In this case specifically, the use of `withStoredKey_Account` in `applyEditingAccountRow` means that the `account` type variable in the signature should really be `Account`. My uneducated guess is that type inference has trouble with what to do with the mentions of `editAccount`.

Long story short, I think you'd want to use `withStoredKeyDict` there instead :)

**Comment by mcapodici** [comment]
@zwilias thanks you got me out of a spot there. Using withStoredKeyDict is what I meant to do. Clever how you figured that without looking at the "calling code".

**Comment by evancz** [comment]
In my development build, I am getting:

```
-- TYPE MISMATCH ------------------------------------------------------ temp.elm

The 3rd argument to `updateItemWithKey` is not what I expect:

125|             ( updateItemWithKey withStoredKey_Account editRowStoredKey updatedAccount accounts
                                                                            ^^^^^^^^^^^^^^
This `updatedAccount` value has type:

    account

But `updateItemWithKey` needs the 3rd argument to be:

    { name : String.String, storedKey : Main.StoredKey }

Hint: Your type annotation uses type variable `account` which means ANY type of
value can flow through, but your code is saying it specifically wants a record.
Maybe change your type annotation to be more specific? Maybe change the code to
be more general?

Read <https://elm-lang.org/hints/0.19.0/type-annotations> for more advice!

Hint: I always figure out the argument types from left to right. If an argument
is acceptable, I assume it is “correct” and move on. So the problem may actually
be in one of the previous arguments!
```

So the recent https://github.com/elm-lang/elm-compiler/commit/4b25980d93ada12236c794f35570e197be82f787 probably fixed this. The error message shown is not perfect, but it's still in development!

Anyway, thank you for the report!
